### DISCUSSION TITLE
Unexpected text is inserted into the buffer after running shell command

### DISCUSSION THREAD

**Discussion by voidcontext** [helix-editor/helix#8606] [Troubleshooting] [answered]
### Summary

After running certain shell commands text like `tty-cmd{"ok": true, "data": "37"}\` is inserted into the current buffer. It seems like `data` sometimes contains the output of the intermediate commands that are piped into another. Unfortunately I don't have a minimally reproducible example, but there's one thing that seems to be common, the scripts that are causing this, all executed `kitty` remote commands under the hood.

### Reproduction Steps

 

I tried this:

1. `hx`

I expected this to happen:

Instead, this happened:

### Helix log

<details><summary>~/.cache/helix/helix.log</summary>

```
please provide a copy of `~/.cache/helix/helix.log` here if possible, you may need to redact some of the lines
```

</details>

### Platform

macOS

### Terminal Emulator

kitty terminal

### Helix Version

helix 23.05

**Comment by the-mikedavis** [comment]
Can you give some examples of commands that do this? Without a reproduction case this is very difficult to debug

**Reply by voidcontext** [reply]
So, I did a bit of testing, and the easiest way to trigger this behaviour is this

1) open kitty terminal (with remote commands enabled)
2) run helix
3) in helix run `:sh kitty @ ls`

This won't give any usual output in the popup (?) window, but instead inserts this into the scratch buffer:

```
lumns\": 192,\n            \"cwd\": \"/Users/gaborpihaj\",\n            \"env\": {},\n            \"foreground_processes\": [\n              {\n                \"cmdline\": [\n                  \"kitten\",\n                  \"@\",\n                  \"ls\"\n                ],\n                \"cwd\": \"/Users/gaborpihaj\",\n                \"pid\": 5857\n              },\n              {\n                \"cmdline\": [\n                  \"/etc/profiles/per-user/gaborpihaj/bin/hx\"\n                ],\n                \"cwd\": \"/Users/gaborpihaj\",\n                \"pid\": 5852\n              }\n            ],\n            \"id\": 1,\n            \"is_active\": true,\n            \"is_focused\": true,\n            \"is_self\": true,\n            \"lines\": 44,\n            \"pid\": 4833,\n            \"title\": \"hx\"\n          }\n        ]\n      }\n    ],\n    \"wm_class\": \"kitty\",\n    \"wm_name\": \"kitty\"\n  }\n]"}\
```

This looks like the partial output of the `kitty @ ls` command. The result is the same even if the `kitty @ ls` command is wrapped in a bash script, or when it's output is redirected into a file:

```bash
❯ cat test-kitty.sh 
#!/usr/bin/env bash

kitty @ ls > kitty.log
❯ export PATH=.:$PATH
❯ hx
```

Then `:sh test-kitty.sh` inserts the same text into the buffer, and `kitty.log` is actually empty.

Another examples
- `:sh kitty @ focus-tab` inserts `tty-cmd{"ok":·true}\`
- `:sh kitty @ set-window-title foobar` doesn't insert anything, only a `Command succeeded` message is displayed.

**Reply by voidcontext** [reply]
The previous tests were using kitty 0.28.1, with the latest version (0.30.1), the behaviour is slightly different:

- `:sh kitty @ ls` still inserts the text, but the output is also displayed
- `:sh test-kitty.sh` behaves the same way

**Comment by kovidgoyal** [comment, marked as answer]
Either setup a unix socket for kitty remote control or make sure helix is not doing any I/O to the tty while kitten @ is running, because kitten @ does I/O to the tty in the absence of a remote control socket.

**Reply by voidcontext** [reply]
Using a unix socket solved the issue, thanks for the help and the explanation!
